crates/bevy_hierarchy/src/child_builder.rs

```diff
@@ -20,16 +95,32 @@ pub struct AddChild {
 
 impl Command for AddChild {
     fn write(self, world: &mut World) {
-        world
-            .entity_mut(self.child)
-            // FIXME: don't erase the previous parent (see #1545)
-            .insert_bundle((Parent(self.parent), PreviousParent(self.parent)));
-        if let Some(mut children) = world.get_mut::<Children>(self.parent) {
-            children.0.push(self.child);
+        let previous = update_parent(world, self.child, self.parent);
+        if let Some(previous) = previous {
+            if previous == self.parent {
+                return;
+            }
+            remove_from_children(world, previous, self.child);
+            if let Some(mut events) = world.get_resource_mut::<Events<HierarchyEvent>>() {
+                events.send(HierarchyEvent::ChildMoved {
+                    child: self.child,
+                    previous_parent: previous,
+                    new_parent: self.parent,
+                });
+            }
+        } else if let Some(mut events) = world.get_resource_mut::<Events<HierarchyEvent>>() {
+            events.send(HierarchyEvent::ChildAdded {
+                child: self.child,
+                parent: self.parent,
+            });
+        }
+        let mut parent = world.entity_mut(self.parent);
+        if let Some(mut children) = parent.get_mut::<Children>() {
+            if !children.contains(&self.child) {
```

Comment:
Is this check necessary? Given that this is now transactional, won't `children` only contain `child` if `child` already has the `Parent` component pointing to `self.parent`? In that case, the `update_parent` call would have returned `Some(self.parent)` and `AddChild` would early return on line 101.

Reply:
Not doing this check could result in duplicate children if the hierarchy is altered directly via `World`. It's highly unlikely given that there's no easy way to make those edits, but it is doable.